File: src/bmsx_cpp/render/vdp/texture_transfer.cpp
#include "render/vdp/texture_transfer.h"

#include "render/gameview.h"
#include "render/texture_manager.h"

namespace bmsx {
namespace {

const TextureParams DEFAULT_TEXTURE_PARAMS{};
TextureManager* g_vdpTextureManager = nullptr;
GameView* g_vdpTextureView = nullptr;

} // namespace

void initializeVdpTextureTransfer(TextureManager& textureManager, GameView& view) {
	g_vdpTextureManager = &textureManager;
	g_vdpTextureView = &view;
}

GPUBackend& vdpTextureBackend() {
	return *g_vdpTextureView->backend();
}

TextureHandle vdpTextureByUri(const std::string& textureKey) {
	return g_vdpTextureManager->getTextureByUri(textureKey);
}

TextureHandle createVdpTextureFromSeed(const std::string& textureKey, const u8* seedPixel, u32 width, u32 height) {
	auto& texmanager = *g_vdpTextureManager;
	texmanager.createTextureFromPixelsSync(textureKey, seedPixel, 1, 1, DEFAULT_TEXTURE_PARAMS);
	TextureHandle handle = vdpTextureByUri(textureKey);
	handle = texmanager.resizeTextureForKey(textureKey, static_cast<i32>(width), static_cast<i32>(height));
	g_vdpTextureView->textures[textureKey] = handle;
	return handle;
}

TextureHandle createVdpTextureFromPixels(const std::string& textureKey, const u8* pixels, u32 width, u32 height) {
	auto& texmanager = *g_vdpTextureManager;
	TextureHandle handle = texmanager.createTextureFromPixelsSync(textureKey, pixels, static_cast<i32>(width), static_cast<i32>(height), DEFAULT_TEXTURE_PARAMS);
	handle = texmanager.resizeTextureForKey(textureKey, static_cast<i32>(width), static_cast<i32>(height));
	vdpTextureBackend().updateTexture(handle, pixels, static_cast<i32>(width), static_cast<i32>(height), DEFAULT_TEXTURE_PARAMS);
	g_vdpTextureView->textures[textureKey] = handle;
	return handle;
}

TextureHandle resizeVdpTextureForKey(const std::string& textureKey, u32 width, u32 height) {
	TextureHandle handle = g_vdpTextureManager->resizeTextureForKey(
		textureKey,
		static_cast<i32>(width),
		static_cast<i32>(height)
	);
	g_vdpTextureView->textures[textureKey] = handle;
	return handle;
}

TextureHandle updateVdpTexturePixels(const std::string& textureKey, const u8* pixels, u32 width, u32 height) {
	TextureHandle handle = resizeVdpTextureForKey(textureKey, width, height);
	vdpTextureBackend().updateTexture(handle, pixels, static_cast<i32>(width), static_cast<i32>(height), DEFAULT_TEXTURE_PARAMS);
	return handle;
}

void updateVdpTextureRegion(const std::string& textureKey, const u8* pixels, i32 width, i32 height, i32 x, i32 y) {
	vdpTextureBackend().updateTextureRegion(vdpTextureByUri(textureKey), pixels, width, height, x, y, DEFAULT_TEXTURE_PARAMS);
}

void swapVdpTextureHandlesByUri(const std::string& textureKeyA, const std::string& textureKeyB) {
	g_vdpTextureManager->swapTextureHandlesByUri(textureKeyA, textureKeyB);
	g_vdpTextureView->textures[textureKeyA] = vdpTextureByUri(textureKeyA);
	g_vdpTextureView->textures[textureKeyB] = vdpTextureByUri(textureKeyB);
}

} // namespace bmsx

File: src/bmsx/render/vdp/texture_transfer.ts
import type { GPUBackend, TextureHandle } from '../backend/backend';
import { DEFAULT_TEXTURE_PARAMS } from '../backend/texture_params';
import type { TextureManager } from '../texture_manager';
import type { GameView } from '../gameview';

let textureManager: TextureManager;
let textureView: GameView;

export function initializeVdpTextureTransfer(manager: TextureManager, view: GameView): void {
	textureManager = manager;
	textureView = view;
}

export function vdpTextureBackend(): GPUBackend {
	return textureView.backend;
}

// disable-next-line single_line_method_pattern -- VDP texture memory owns texture-manager lookup; callers should not reach into the manager.
export function vdpTextureByUri(textureKey: string): TextureHandle {
	return textureManager.getTextureByUri(textureKey);
}

export function createVdpTextureFromSeed(textureKey: string, seedPixel: Uint8Array, width: number, height: number): TextureHandle {
	textureManager.createTextureFromPixelsSync(textureKey, seedPixel, 1, 1);
	const handle = textureManager.resizeTextureForKey(textureKey, width, height);
	textureView.textures[textureKey] = handle;
	return handle;
}

export function createVdpTextureFromPixels(textureKey: string, pixels: Uint8Array, width: number, height: number): TextureHandle {
	textureManager.createTextureFromPixelsSync(textureKey, pixels, width, height);
	const handle = textureManager.resizeTextureForKey(textureKey, width, height);
	vdpTextureBackend().updateTexture(handle, pixels, width, height, DEFAULT_TEXTURE_PARAMS);
	textureView.textures[textureKey] = handle;
	return handle;
}

export function resizeVdpTextureForKey(textureKey: string, width: number, height: number): TextureHandle {
	const handle = textureManager.resizeTextureForKey(textureKey, width, height);
	textureView.textures[textureKey] = handle;
	return handle;
}

export function updateVdpTexturePixels(textureKey: string, pixels: Uint8Array, width: number, height: number): TextureHandle {
	const handle = resizeVdpTextureForKey(textureKey, width, height);
	vdpTextureBackend().updateTexture(handle, pixels, width, height, DEFAULT_TEXTURE_PARAMS);
	return handle;
}

export function updateVdpTextureRegion(textureKey: string, pixels: Uint8Array, width: number, height: number, x: number, y: number): void {
	vdpTextureBackend().updateTextureRegion(vdpTextureByUri(textureKey), pixels, width, height, x, y, DEFAULT_TEXTURE_PARAMS);
}

export function swapVdpTextureHandlesByUri(textureKeyA: string, textureKeyB: string): void {
	textureManager.swapTextureHandlesByUri(textureKeyA, textureKeyB);
	textureView.textures[textureKeyA] = vdpTextureByUri(textureKeyA);
	textureView.textures[textureKeyB] = vdpTextureByUri(textureKeyB);
}
